handlers/changes.js

```diff
@@ -16,112 +22,238 @@ var getViewKeys = function(req, userCtx, callback) {
-        return callback(err);
-      }
-      var ids = _.pluck(viewResult.rows, 'id');
-      ids.push('org.couchdb.user:' + userCtx.name);
-      callback(null, ids);
-    });
+    var ids = _.pluck(viewResult.rows, 'id');
+    ids.push('org.couchdb.user:' + feed.userCtx.name);
+    feed.validatedIds = ids;
+    callback();
+  });
+};
+
+var bindRequestedIds = function(feed, callback) {
+  var ids = [];
+  if (feed.req.body && feed.req.body.doc_ids) {
+    // POST request
+    ids = feed.req.body.doc_ids;
+  } else if (feed.req.query && feed.req.query.doc_ids) {
+    // GET request
+    try {
+      ids = JSON.parse(feed.req.query.doc_ids);
+    } catch(e) {
+      return callback({ code: 400, message: 'Invalid doc_ids param' });
+    }
+  }
+  feed.requestedIds = ids;
+  return callback();
+};
+
+var defibrillator = function(feed) {
+  if (feed.req.query && feed.req.query.heartbeat) {
+    feed.heartbeat = setInterval(function() {
+      feed.res.write('\n');
+    }, feed.req.query.heartbeat);
+  }
+};
+
+var prepareResponse = function(feed, changes) {
+  var rejected = _.reject(changes.results, function(change) {
+    return change.deleted || _.contains(feed.validatedIds, change.id);
   });
+  if (rejected.length) {
+    console.error('User attempting to replicate these docs without permission: ', _.pluck(rejected, 'id'));
+    return feed.res.write(error(403, 'Forbidden'));
+  }
+  feed.res.write(JSON.stringify(changes));
 };
 
-var getChanges = function(req, ids, callback) {
-  var params = _.pick(req.query, 'timeout', 'style', 'heartbeat', 'since', 'feed', 'limit', 'filter');
+var getChanges = function(feed) {
   // we cannot call 'changes' in nano because it only uses GET requests and
   // our query string might be too long for GET
-  db.request({
+  feed.changesReq = db.request({
     db: db.settings.db,
     path: '_changes',
-    qs: params,
-    body: { doc_ids: ids },
+    qs:  _.pick(feed.req.query, 'timeout', 'style', 'heartbeat', 'since', 'feed', 'limit', 'filter'),
+    body: { doc_ids: _.union(feed.requestedIds, feed.validatedIds) },
     method: 'POST'
-  }, callback);
+  }, function(err, changes) {
+    if (feed.heartbeat) {
+      clearInterval(feed.heartbeat);
+    }
+    if (err) {
+      feed.res.write(error(503, 'Error processing your changes'));
+    } else {
+      prepareResponse(feed, changes);
+    }
+    feed.res.end();
+    continuousFeeds = _.reject(continuousFeeds, function(feed) {
+      return feed.res.finished;
+    });
+  });
 };
 
-var prepareResponse = function(req, res, changes, verifiedIds) {
-  var rejected = _.reject(changes.results, function(change) {
-    return change.deleted || _.contains(verifiedIds, change.id);
+var initFeed = function(feed, callback) {
+  bindRequestedIds(feed, function(err) {
+    if (err) {
+      return callback(err);
+    }
+    bindViewKeys(feed, function(err) {
+      if (err) {
+        return callback(err);
+      }
+      bindValidatedDocIds(feed, callback);
+    });
```

Comment:
The callbacks vs promises argument rages on. In most cases they are completely equivalent. There is one special case where callbacks are required (multiple invocations) and one case when promises are much nicer (asynchronous initialisation), but other than that it's personal preference. My personal preference is to use promises.

However, it's much nicer to use one or the other as much as possible within one codebase because switching between is buggy and ugly (buggly?). In API we use callbacks everywhere so I stuck with it here for consistency. I'm open to the idea of refactoring API to use promises.